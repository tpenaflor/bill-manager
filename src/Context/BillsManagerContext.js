import React from 'react'

const BillsManagerContext = React.createContext()

export default BillsManagerContext