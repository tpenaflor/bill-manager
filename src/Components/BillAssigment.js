import React, {useState} from 'react'
import {useBillsManagerContext} from '../Context/BillsManagerContext'

import {participantBill} from '../Selector/BillSelector'
import {currency} from '../tools/formatter'

const BillAssignment = ({part, bill}) => {
    const {bills, billDispatch} = useBillsManagerContext()
    const [isActive, setActive] = useState(!!bill.allocation.participant.find(({id})=>id===part.id))
    const [unitCount, setUnitCount] = useState(bill.allocation.allocated)

    const setUnitAllocation = (e)=> {
        setUnitCount(e.target.value)
        
        const participant = {
            id : part.id,
            count : parseInt(e.target.value)
        }

        billDispatch({
            type : "SET_ALLOCATION",
            participant,
            id : bill.id
        })
    }

    const handleCheckBox = () => {
        setActive(!isActive)
        const participant = {
            id : part.id,
            count : 1
        }

        const offset = !isActive ? 1 : -1
        const type = !isActive ? "ADD_PARTICIPANT_TO_BILL" : "DEL_PARTICIPANT_FROM_BILL"

        billDispatch({
            type,
            participant,
            unitCount : bill.allocation.participant.length + offset,
            id : bill.id
        })
    }

    return (
        <div>
            {bill.splitOpt === 'even' ? (
                <input type="checkbox" 
                    checked={isActive} 
                    onChange={handleCheckBox}  />
            ) :
            (
                 <select id='unitCount'
                    value={unitCount}
                    onChange={setUnitAllocation}>
                     {
                         [...Array(parseInt(bill.allocation.unitCount) + 1)].map((e, i)=>(
                             <option key={i}>{i}</option>
                             ))
                     }
                 </select>
            )}
            
            <p>{`${part.name} ${bill.desc} ${currency(participantBill(part.id, bill.id, bills))}`}</p> 
        </div>
    )
}


export default BillAssignment