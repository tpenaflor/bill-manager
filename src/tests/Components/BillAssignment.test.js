import React from 'react'
import {shallow} from 'enzyme'
import { shallowToJson  } from 'enzyme-to-json';

import BillAssigment from '../../Components/BillAssigment'
import * as BillsManagerContext from '../../Context/BillsManagerContext'
import {billFixture} from '../Fixtures/BillFixture'
import {participantFixture} from '../Fixtures/ParticipantFixture'

let billDispatch, wrapper

beforeEach( ()=> {
    billDispatch = jest.fn()

    const contextValues = {
        billDispatch,
        bills : billFixture
    }

    jest.spyOn(BillsManagerContext, 'useBillsManagerContext')
        .mockImplementation(() => contextValues)
})

afterEach(()=>{
    expect(shallowToJson(wrapper)).toMatchSnapshot()
})

test('set bill allocation to participant', ()=> {
    wrapper = shallow(<BillAssigment part={participantFixture[0]} bill={billFixture[0]}/>)
    expect(shallowToJson(wrapper)).toMatchSnapshot()

    const value = 3
    wrapper.find('select').simulate('change', {target:{value}})

    expect(billDispatch).toHaveBeenLastCalledWith({
        type : "SET_ALLOCATION",
        participant : {
            id : participantFixture[0].id,
            count : value
        },
        id : billFixture[0].id
    })
})

test('tick participant checkbox', ()=> {
    wrapper = shallow(<BillAssigment part={participantFixture[1]} bill={billFixture[2]}/>)
    expect(shallowToJson(wrapper)).toMatchSnapshot()

    wrapper.find('input').simulate('change')

    expect(billDispatch).toHaveBeenLastCalledWith({
        type : "ADD_PARTICIPANT_TO_BILL",
        participant:{
            id : participantFixture[1].id,
            count : 1
        },
        unitCount : billFixture[2].allocation.participant.length + 1,
        id : billFixture[2].id
    })
})


test('untick participant checkbox', ()=> {
    wrapper = shallow(<BillAssigment part={participantFixture[0]} bill={billFixture[2]}/>)
    expect(shallowToJson(wrapper)).toMatchSnapshot()

    wrapper.find('input').simulate('change')

    expect(billDispatch).toHaveBeenLastCalledWith({
        type : "DEL_PARTICIPANT_FROM_BILL",
        participant:{
            id : participantFixture[0].id,
            count : 1
        },
        unitCount : billFixture[2].allocation.participant.length -1,
        id : billFixture[2].id
    })
})