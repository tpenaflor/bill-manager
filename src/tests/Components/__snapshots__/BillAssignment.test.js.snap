// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`set bill allocation to participant 1`] = `
<div>
  <select
    id="unitCount"
    onChange={[Function]}
  >
    <option
      key="0"
    >
      0
    </option>
    <option
      key="1"
    >
      1
    </option>
    <option
      key="2"
    >
      2
    </option>
    <option
      key="3"
    >
      3
    </option>
    <option
      key="4"
    >
      4
    </option>
    <option
      key="5"
    >
      5
    </option>
    <option
      key="6"
    >
      6
    </option>
    <option
      key="7"
    >
      7
    </option>
    <option
      key="8"
    >
      8
    </option>
    <option
      key="9"
    >
      9
    </option>
    <option
      key="10"
    >
      10
    </option>
  </select>
  <p>
    user1 beer $0.00
  </p>
</div>
`;

exports[`set bill allocation to participant 2`] = `
<div>
  <select
    id="unitCount"
    onChange={[Function]}
    value={3}
  >
    <option
      key="0"
    >
      0
    </option>
    <option
      key="1"
    >
      1
    </option>
    <option
      key="2"
    >
      2
    </option>
    <option
      key="3"
    >
      3
    </option>
    <option
      key="4"
    >
      4
    </option>
    <option
      key="5"
    >
      5
    </option>
    <option
      key="6"
    >
      6
    </option>
    <option
      key="7"
    >
      7
    </option>
    <option
      key="8"
    >
      8
    </option>
    <option
      key="9"
    >
      9
    </option>
    <option
      key="10"
    >
      10
    </option>
  </select>
  <p>
    user1 beer $0.00
  </p>
</div>
`;

exports[`tick participant checkbox 1`] = `
<div>
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />
  <p>
    user2 water $0.00
  </p>
</div>
`;

exports[`tick participant checkbox 2`] = `
<div>
  <input
    checked={true}
    onChange={[Function]}
    type="checkbox"
  />
  <p>
    user2 water $0.00
  </p>
</div>
`;

exports[`untick participant checkbox 1`] = `
<div>
  <input
    checked={true}
    onChange={[Function]}
    type="checkbox"
  />
  <p>
    user1 water $200.00
  </p>
</div>
`;

exports[`untick participant checkbox 2`] = `
<div>
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />
  <p>
    user1 water $200.00
  </p>
</div>
`;
